### ISSUE TITLE
`std::logic_error (Stream ends with an active range tombstone: ...)` when performing paged queries on `master`

### ISSUE THREAD

**Issue by kbr-** [scylladb/scylla#9907]
At commit 134601a15e9fc248e01d665061ffe5904f66362e the dtest `reversed_queries_test.py:TestReversedQueriesWithOverlappingRangeTombstones.test_reversed_query_with_overlapping_range_tombstones`
(https://github.com/scylladb/scylla-dtest/blob/branch-4.6/reversed_queries_test.py#L472)
fails

```
======================================================================
FAIL: test_reversed_query_with_overlapping_range_tombstones (reversed_queries_test.TestReversedQueriesWithOverlappingRangeTombstones)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/kbraun/dev/scylla-dtest/dtest.py", line 1148, in tearDown
    self.cleanUpCluster()
  File "/home/kbraun/dev/scylla-dtest/dtest.py", line 1202, in cleanUpCluster
    raise AssertionError('Unexpected errors found: {}'.format(found_errors))
AssertionError: Unexpected errors found: [('node1', ['ERROR 2022-01-12 18:44:38,592 [shard 0] storage_proxy - Exception when communicating with 127.0.20.1, to read from ks.t: std::logic_error (Stream ends with an active range tombstone: {range_tombstone_change: pos={position: clustered,ckp{000400030d36},1}, {tombstone: timestamp=1642009478587985, deletion_time=1642009478}})'])]
-------------------- >> begin captured logging << --------------------
dtest: DEBUG: going to run tests sequentially
dtest: DEBUG: using the RandomClusterIdAllocator
dtest: DEBUG: reversed_queries_test.py:TestReversedQueriesWithOverlappingRangeTombstones.test_reversed_query_with_overlapping_range_tombstones - cluster ccm directory: /home/kbraun/.dtest/dtest-zcw93cny
dtest: DEBUG: reversed_queries_test.py:TestReversedQueriesWithOverlappingRangeTombstones.test_reversed_query_with_overlapping_range_tombstones - Starting Scylla cluster from directory /home/kbraun/dev/scylla-dtest/../scylla/build/dev
ccm: DEBUG: Started cluster 'test' version 4.7.dev installed in /home/kbraun/dev/scylla
dtest: DEBUG: reversed_queries_test.py:TestReversedQueriesWithOverlappingRangeTombstones.test_reversed_query_with_overlapping_range_tombstones - Allocated cluster ID 20: /home/kbraun/.dtest/dtest-zcw93cny
dtest: DEBUG: reversed_queries_test.py:TestReversedQueriesWithOverlappingRangeTombstones.test_reversed_query_with_overlapping_range_tombstones - Scylla mode is 'dev'
dtest: DEBUG: reversed_queries_test.py:TestReversedQueriesWithOverlappingRangeTombstones.test_reversed_query_with_overlapping_range_tombstones - Cluster *_request_timeout_in_ms=20000, range_request_timeout_in_ms=60000, cql request_timeout=120
dtest: DEBUG: reversed_queries_test.py:TestReversedQueriesWithOverlappingRangeTombstones.test_reversed_query_with_overlapping_range_tombstones - configuring skip_wait_for_gossip_to_settle=0 for single_node test
dtest: DEBUG: reversed_queries_test.py:TestReversedQueriesWithOverlappingRangeTombstones.test_reversed_query_with_overlapping_range_tombstones - Set up a cluster
ccm: DEBUG: node1: Starting scylla: args=['/home/kbraun/.dtest/dtest-zcw93cny/test/node1/bin/scylla', '--options-file', '/home/kbraun/.dtest/dtest-zcw93cny/test/node1/conf/scylla.yaml', '--log-to-stdout', '1', '--api-address', '127.0.20.1', '--collectd-hostname', 'localhost.localdomain.node1', '--abort-on-seastar-bad-alloc', '--abort-on-lsa-bad-alloc', '1', '--developer-mode', 'true', '--smp', '1', '--memory', '512M', '--default-log-level', 'info', '--collectd', '0', '--overprovisioned', '--prometheus-address', '127.0.20.1', '--unsafe-bypass-fsync', '1', '--kernel-page-cache', '1'] wait_other_notice=False wait_for_binary_proto=False
dtest: DEBUG: reversed_queries_test.py:TestReversedQueriesWithOverlappingRangeTombstones.test_reversed_query_with_overlapping_range_tombstones - Set up schema
dtest: DEBUG: reversed_queries_test.py:TestReversedQueriesWithOverlappingRangeTombstones.test_reversed_query_with_overlapping_range_tombstones - Generate range tombstones
dtest: DEBUG: reversed_queries_test.py:TestReversedQueriesWithOverlappingRangeTombstones.test_reversed_query_with_overlapping_range_tombstones - Insert a row near the end of the range tombstones
dtest: DEBUG: reversed_queries_test.py:TestReversedQueriesWithOverlappingRangeTombstones.test_reversed_query_with_overlapping_range_tombstones - Select the row and check result
dtest: DEBUG: reversed_queries_test.py:TestReversedQueriesWithOverlappingRangeTombstones.test_reversed_query_with_overlapping_range_tombstones - Test failed with errors: [(<reversed_queries_test.TestReversedQueriesWithOverlappingRangeTombstones testMethod=test_reversed_query_with_overlapping_range_tombstones>, (<class 'cassandra.ReadFailure'>, ReadFailure('Error from server: code=1300 [Replica(s) failed to execute read] message="Operation failed for ks.t - received 0 responses and 1 failures from 1 CL=ONE." info={\'consistency\': \'ONE\', \'required_responses\': 1, \'received_responses\': 0, \'failures\': 1}'), <traceback object at 0x7fad685e3e80>))]
dtest: DEBUG: reversed_queries_test.py:TestReversedQueriesWithOverlappingRangeTombstones.test_reversed_query_with_overlapping_range_tombstones - Stopping cluster
dtest: DEBUG: reversed_queries_test.py:TestReversedQueriesWithOverlappingRangeTombstones.test_reversed_query_with_overlapping_range_tombstones - stopping ccm cluster test at: /home/kbraun/.dtest/dtest-zcw93cny
dtest: DEBUG: reversed_queries_test.py:TestReversedQueriesWithOverlappingRangeTombstones.test_reversed_query_with_overlapping_range_tombstones - Freeing cluster ID 20: link /home/kbraun/.dtest/20
--------------------- >> end captured logging << ---------------------

```

A regression must have happened somewhere in this patchset, since the test passes at the previous commit (4118f2d8bee9e945c9febc297b4bab60d9749320)

cc @denesb

**Comment by kbr-** [comment]
Huh, apparently this has nothing to do with reversed queries. It seems I can reproduce it with forward queries as well

**Comment by kbr-** [comment]
Repro:

```python
with Cluster(['127.0.0.1']) as cluster:
    session = cluster.connect()
    session.execute("create keyspace if not exists ks with replication = {'class': 'SimpleStrategy', 'replication_factor': 1};")
    session.execute("create table if not exists ks.t (pk int, ck int, v int, primary key (pk, ck));")
    query = session.prepare("insert into ks.t (pk, ck, v) values (0, ?, 0)")

    expected_keys = []
    rt_strs = []

    def insert_random_range_tombstone():
        rt_start = random.randint(0, 200)
        rt_start_inclusive = random.choice([True, False])
        rt_end = rt_start + random.randint(0, 10)
        rt_end_inclusive = random.choice([True, False])

        rt_left_cmp = '>=' if rt_start_inclusive else '>'
        rt_right_cmp = '<=' if rt_end_inclusive else '<'
        session.execute(f'delete from ks.t where pk = 0 and ck {rt_left_cmp} {rt_start}'
                        f' and ck {rt_right_cmp} {rt_end}')

        rt_left_str = '[' if rt_start_inclusive else '('
        rt_right_str = ']' if rt_end_inclusive else ')'
        rt_strs.append(f'{rt_left_str}{rt_start}, {rt_end}{rt_right_str}')

        def filter_key(k):
            return not ((k > rt_start or (rt_start_inclusive and k == rt_start))
                        and (k < rt_end or (rt_end_inclusive and k == rt_end)))

        return filter_key

    # Put some data in memtable and some in sstable
    print('Insert data')
    for key in range(100):
        session.execute(query, (key,))
        expected_keys.append(key)
    for _ in range(10):
        filter_key = insert_random_range_tombstone()
        expected_keys = [k for k in expected_keys if filter_key(k)]
    print('Flush')
    requests.post(f'http://127.0.0.1:10000/storage_service/keyspace_flush/ks', params={'cf' : 't'})
    print('Insert more data')
    for key in range(100, 200):
        session.execute(query, (key,))
        expected_keys.append(key)
    for _ in range(10):
        filter_key = insert_random_range_tombstone()
        expected_keys = [k for k in expected_keys if filter_key(k)]

    expected_keys = list(expected_keys)
    print(f'Expected keys: {expected_keys}')
    print(f'Inserted range tombstones: {rt_strs}')

    query = session.prepare("select * from ks.t where pk = 0 bypass cache")
    query.fetch_size = 10
    fetched_keys = [r.ck for r in session.execute(query)]

    print(f'Fetched keys: {fetched_keys}')
    assert expected_keys == fetched_keys

```

apparently there's some problem with paging

**Comment by denesb** [comment]
I reproduced this locally as well. I'm investigating.

**Comment by jm-ds** [comment]
We're running 5.2 on ubuntu. We upgraded our entire cluster last week from 4.6(incremental upgrades). Now we're seeing thousands of these errors in the logs and one of the nodes in our cluster spontaneously shutdown yesterday. 

`[shard  1] view_update_generator - Processing threatrix failed for table component_revised_license:std::logic_error (Stream ends with an active range tombstone: {range_tombstone_change: pos={position: clustered,ckp{0013302e342e302d6e6578742e3834666462343138000765706c2d322e30},1}, {tombstone: timestamp=1685198768252762, deletion_time=1685198768}}). Will retry...`

**Comment by michoecho** [comment]
> Could this be due to https://github.com/scylladb/scylladb/pull/14502 @michoecho ?

You mean the issue (#14503), not the fix (#14502)?

Yes, it could.

**Comment by jm-ds** [comment]
@tgrabiec @michoecho I'm not clear on what the fix is for this issue. The node exhibiting this issue is at a constant 50% CPU utilization and it's killing our platform. A fix path would be greatly appreciated.

**Comment by michoecho** [comment]
> @tgrabiec @michoecho I'm not clear on what the fix is for this issue. The node exhibiting this issue is at a constant 50% CPU utilization and it's killing our platform. A fix path would be greatly appreciated.

We suspect you've hit issue https://github.com/scylladb/scylladb/issues/14503. That issue has been fixed by https://github.com/scylladb/scylladb/pull/14502.

The fix didn't make it to an official patch release yet. But you could install a recent nightly build of the 5.2 branch, which contains the fix.
Nightly builds don't go through a full QA cycle, but stable branches only receive bugfixes, so the risk of regressions in a stable branch is relatively low. But even though it's very rare, unfortunately it does happen sometimes; you have been warned.
The latest nightly build is https://downloads.scylladb.com/unstable/scylla/branch-5.2/build-packages/2023-07-20T16%3A56%3A07Z/deb/
(Here's the list of changes with respect to 5.2.4: https://github.com/scylladb/scylladb/compare/scylla-5.2.4...986491447b981bdb88e8076a30b2dd67af7d7d32)

Some workarounds for the issue are also possible. The bug is triggered during view updates after repair/streaming. So a possible workaround is to stop any repair or streaming, stop the ill node, back up the contents of the `staging` directory for the affected table (and leave the `staging` directory clear), start the node. Then wait for a release with the fix, stop the node, bring back the `staging` sstables, start the node, and only then resume repair/streaming. But since no repair is possible in the meantime, care has to be taken to prevent data resurrection. And I don't know this part of the system that well, I might be forgetting something in my description. You shouldn't attempt the workaround until someone more knowledgeable than me in this area reviews it.

**Comment by denesb** [comment]
> So a possible workaround is to stop any repair or streaming, stop the ill node, back up the contents of the `staging` directory for the affected table (and leave the `staging` directory clear), start the node.

This results in effectively loosing the data that is in these staging directories, while they are moved aside.
An alternative solution is to stop scylla, move all staging sstables into the `upload` directory, then start scylla. The sstables will be merged into the main set without building view updates for them. This can make the views become out-of-sync and possibly will require rebuilding all views once a version with a fix is available.

**Comment by jm-ds** [comment]
@denesb @michoecho This issue was resolved for us by upgrading to scylla-5.2.5. Thank you for all of your time and assistance.
